File: problems/Java/46. Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        bt(res, new ArrayList<>(), nums, new boolean[nums.length]);
        return res;
    }

    private void bt(List<List<Integer>> res, List<Integer> path, int[] nums, boolean[] used) {
        if (path.size() == nums.length) {
            res.add(new ArrayList<>(path));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (!used[i]) {
                used[i] = true;
                path.add(nums[i]);
                bt(res, path, nums, used);
                path.remove(path.size() - 1);
                used[i] = false;
            }
        }
    }
}

File: problems/C++/46. Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        vector<bool> used(nums.size(), false);
        bt(res, path, nums, used);
        return res;
    }

private:
    void bt(vector<vector<int>>& res, vector<int>& path, vector<int>& nums, vector<bool>& used) {
        if (path.size() == nums.size()) {
            res.push_back(path);
            return;
        }

        for (int i = 0; i < nums.size(); ++i) {
            if (!used[i]) {
                used[i] = true;
                path.push_back(nums[i]);
                bt(res, path, nums, used);
                path.pop_back();
                used[i] = false;
            }
        }
    }
};